lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -558,6 +558,220 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
+        // perform collapse and reshape operations to match the ranks of data
+        // and indices(making sure the semantics of the onnx.gather_nd are
+        // preserved), perform torch.gather operation, later expand the gather
+        // result to match onnx.gather_nd output. For example, assuming indices
+        // is of shape (4, 5, 3, 2), data is (4, 10, 11, 7, 4) and
+        // batch_dims(`b`)=1. Firstly, modify indices to 1-D indexing as the
+        // torch.gather op supports only single dimensional indexing. (this
+        // algorithm would have been simpler if we can get a torch op that
+        // supports indexing at multiple dimensions simultaneously). 1-D indexed
+        // indices will be of shape (4, 5, 3, 1), now materialize it to
+        // `r-b-indices_shape[-1]` dimension of data i.e. reshaping it to the
+        // shape (4, 5, 3, 1, 1). Next step is to flatten the indices and data
+        // to (4, 15, 1, 1) and (4, 110, 7, 4) shapes respectively and then
+        // perform the torch.gather operation. Post the gather operation,
+        // unflatten the indices dimensions of result to (4, 5, 3, 1, 1) and
+        // then expand it to get the final output of shape (4, 5, 3, 7, 4).
+
+        // step3. Convert indices_shape[-1] dimensional indexing to 1D indexing.
+        Value constZero = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(0));
+        Value constOne = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(1));
+        Value sliceDim = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(indicesRank - 1));
+        SmallVector<int64_t> indicesSliceShape(indicesShapeMinusOne);
+        indicesSliceShape.push_back(1);
+        auto indicesSliceTy = rewriter.getType<Torch::ValueTensorType>(
+            indicesSliceShape, indicesTy.getOptionalDtype());
+
+        Value start = constZero;
+        Value updatedIndices;
+        for (int64_t i = 0; i < indicesLastDim; ++i) {
+          Value end = rewriter.create<Torch::ConstantIntOp>(
+              loc, rewriter.getI64IntegerAttr(i + 1));
+          Value indicesSlice = rewriter.create<Torch::AtenSliceTensorOp>(
+              loc, indicesSliceTy, indices, sliceDim, start, end,
+              /*step=*/constOne);
+          // Apply bounds checking on the indices slice.
+          auto boolTy = rewriter.getType<Torch::ValueTensorType>(
+              indicesSliceShape, rewriter.getI1Type());
+          Value lt = rewriter.create<Torch::AtenLtScalarOp>(
+              loc, boolTy, indicesSlice, constZero);
+          Value add = rewriter.create<Torch::AtenAddScalarOp>(
+              loc, indicesSliceTy, indicesSlice, dataDims[batchDimCount + i],
+              /*alpha=*/constOne);
+          indicesSlice = rewriter.create<Torch::AtenWhereSelfOp>(
+              loc, indicesSliceTy, lt, add, indicesSlice);
+          if (i == 0)
```

Comment:
Avoid if/else. In this case

updatedIndices = indicesSlice;
if (i != 0)
  updateIndices = rewriter.create...

Reply:
updatedIndices should be initialized with indicesSlice only at i=0. 
I can do something like this:
if(i==0){
  updatedIndices = indicesSlice;
  continue;}
updatedIndices = ....